internal/pgadmin/runner.go

```diff
@@ -0,0 +1,209 @@
+
+/*
+ Copyright 2020 Crunchy Data Solutions, Inc.
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/crunchydata/postgres-operator/kubeapi"
+
+	log "github.com/sirupsen/logrus"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+const (
+	defaultPath = "/var/lib/pgadmin/pgadmin4.db"
+	maxRetries  = 10
+)
+
+// queryRunner provides a helper for performing queries against the pgadmin
+// sqlite database via Kubernetes Exec functionality
+type queryRunner struct {
+	BackoffPolicy Backoff
+	Namespace     string
+	Path          string
+	Pod           v1.Pod
+
+	clientset *kubernetes.Clientset
+	apicfg    *rest.Config
+	secSalt   string // Cached value of the database-specific security salt
+	separator string // Field separator for multi-field queries
+	ready     bool   // Flagged true once db has been set up
+}
+
+// NewQueryRunner creates a query runner instance with the configuration
+// necessary to exec into the named pod in the provided namespace
+func NewQueryRunner(clientset *kubernetes.Clientset, apic *rest.Config, pod v1.Pod) *queryRunner {
+	qr := &queryRunner{
+		Namespace: pod.ObjectMeta.Namespace,
+		Path:      defaultPath,
+		Pod:       pod,
+		apicfg:    apic,
+		clientset: clientset,
+		separator: ",",
+	}
+
+	// Set up a default policy as an 'intelligent default', creators can
+	// override, naturally - default will hit max at n == 10
+	qr.BackoffPolicy = ExponentialBackoffPolicy{
+		Base:       35 * time.Millisecond,
+		JitterMode: JitterSmall,
+		Maximum:    2 * time.Second,
+		Ratio:      1.5,
+	}
+
+	return qr
+}
+
+// EnsureReady waits until the database both exists and has content -
+// determined by entires in the user table or the timeout has occurred
+func (qr *queryRunner) EnsureReady() error {
+	// Use cached status to avoid repeated querying
+	if qr.ready {
+		return nil
+	}
+
+	cmd := []string{
+		"sqlite3",
+		qr.Path,
+		"SELECT email FROM user WHERE id=1",
+	}
+
+	// short-fuse test, otherwise minimum wait is tick time
+	stdout, _, err := kubeapi.ExecToPodThroughAPI(qr.apicfg, qr.clientset,
+		cmd, qr.Pod.Spec.Containers[0].Name, qr.Pod.Name, qr.Namespace, nil)
+	if len(strings.TrimSpace(stdout)) > 0 && err == nil {
+		return nil
+	}
+
+	// Use policy that is roughly 90s total timeout
+	backoff := ExponentialBackoffPolicy{
+		Base:       1200 * time.Millisecond,
+		Ratio:      1.25,
+		JitterMode: JitterFull,
+		Maximum:    90,
```

Comment:
Any reason why we don't just raise the maximum to 300 or the like?

Reply:
The Maximum it's referring to is time for a **single** attempt, it caps the exponential growth curve in case someone bumps up maximum retries.

So for this specific case, you're already looking at 20 attempts where the last one takes a full 90s, but the sum total of all events is 501s

Reply:
:+1: